// loaders.gl/dist/es5/gltf-loader/gltf-loader.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.parseTextGLTF = parseTextGLTF;
exports.parseBinaryGLTF = parseBinaryGLTF;
exports.default = void 0;

var _glbParser = _interopRequireDefault(require("../glb-loader/glb-parser"));

var _gltfParser = _interopRequireDefault(require("./gltf-parser"));

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

// Binary container format for glTF
function parseTextGLTF(json) {
  var options = arguments.length > 1 && arguments[1] !== undefined ? arguments[1] : {};
  return new _gltfParser.default(json).parse(options);
}

function parseBinaryGLTF(glbArrayBuffer) {
  var options = arguments.length > 1 && arguments[1] !== undefined ? arguments[1] : {};

  var _parseWithMetadata = new _glbParser.default(glbArrayBuffer).parseWithMetadata(options),
      json = _parseWithMetadata.json,
      arrayBuffer = _parseWithMetadata.arrayBuffer;

  return new _gltfParser.default(json, arrayBuffer).parse(options);
}

var _default = {
  name: 'glTF',
  extension: 'gltf',
  parseText: parseTextGLTF,
  parseBinary: parseBinaryGLTF
};
exports.default = _default;
