// loaders.gl/dist/es5/common/loader-utils/index.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
Object.defineProperty(exports, "loadUri", {
  enumerable: true,
  get: function get() {
    return _loadUri.loadUri;
  }
});
Object.defineProperty(exports, "TextDecoder", {
  enumerable: true,
  get: function get() {
    return _textDecoder.default;
  }
});
Object.defineProperty(exports, "TextEncoder", {
  enumerable: true,
  get: function get() {
    return _textEncoder.default;
  }
});
Object.defineProperty(exports, "getImageSize", {
  enumerable: true,
  get: function get() {
    return _getImageSize.getImageSize;
  }
});
Object.defineProperty(exports, "padTo4Bytes", {
  enumerable: true,
  get: function get() {
    return _arrayUtils.padTo4Bytes;
  }
});
Object.defineProperty(exports, "copyArrayBuffer", {
  enumerable: true,
  get: function get() {
    return _arrayUtils.copyArrayBuffer;
  }
});
Object.defineProperty(exports, "toBuffer", {
  enumerable: true,
  get: function get() {
    return _binaryUtils.toBuffer;
  }
});
Object.defineProperty(exports, "toArrayBuffer", {
  enumerable: true,
  get: function get() {
    return _binaryUtils.toArrayBuffer;
  }
});
Object.defineProperty(exports, "flattenToTypedArray", {
  enumerable: true,
  get: function get() {
    return _flatten.flattenToTypedArray;
  }
});
Object.defineProperty(exports, "toLowPrecision", {
  enumerable: true,
  get: function get() {
    return _toLowPrecision.toLowPrecision;
  }
});

var _loadUri = require("./load-uri.js");

var _textDecoder = _interopRequireDefault(require("./text-decoder"));

var _textEncoder = _interopRequireDefault(require("./text-encoder"));

var _getImageSize = require("./get-image-size");

var _arrayUtils = require("./array-utils");

var _binaryUtils = require("./binary-utils");

var _flatten = require("./flatten");

var _toLowPrecision = require("./to-low-precision");

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }
