// loaders.gl/dist/es5/kml-loader/kml-loader.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _xmlLoader = _interopRequireDefault(require("../formats/xml-loader/xml-loader"));

var _kmlParser = _interopRequireDefault(require("./kml-parser"));

var _kmlNormalizer = _interopRequireDefault(require("./kml-normalizer"));

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

var KML_HEADER = "<?xml version=\"1.0\" encoding=\"UTF-8\"?>\n<kml xmlns=\"http://www.opengis.net/kml/2.2\">\n";
var DEFAULT_KML_OPTIONS = {
  normalize: true
};

function testText(text) {
  return text.startsWith(KML_HEADER);
}

function parseText(text) {
  var options = arguments.length > 1 && arguments[1] !== undefined ? arguments[1] : DEFAULT_KML_OPTIONS;

  var xml = _xmlLoader.default.parseText(text);

  var kmlLoader = new _kmlParser.default(xml);
  var kml = kmlLoader.parse();
  return options.normalize ? (0, _kmlNormalizer.default)(kml) : kml;
}

var _default = {
  name: 'KML',
  extension: 'kml',
  supported: _xmlLoader.default.supported,
  testText: testText,
  parseText: parseText,
  browserOnly: true,
  worker: false
};
exports.default = _default;
